In [138]:
import numpy as np
import heapq

folder = 'C:/Users/Jemandes/Documents/Dropbox/Leiden U/Neural Networks/data/'

In [107]:
train_file_raw = open(folder + 'test_in.csv').readlines()
train_values = open(folder + 'test_out.csv').readlines()
train_file = np.zeros((len(train_file_raw), 257))
train_list = np.zeros((len(train_values), 1))

Digits = np.zeros((10, 257))
Digits[:,0] = np.linspace(0, 9, 10)
dist = np.zeros((10,10))

In [108]:
# make list of outputs
for i, im in enumerate(train_values):
    train_list[i,0] = im
# make combined array: first number is output
for i, image in enumerate(train_file_raw):
    val = image.split(',')
    train_file[i,0] = train_list[i,0]
    for j, el in enumerate( train_file[i,:][1:] ):
        train_file[i,j+1] = val[j]

In [114]:
print (train_file[0,:])

[ 0.    -1.    -1.    -1.    -1.    -1.    -0.966 -0.237  0.847  0.393
 -0.839 -1.    -1.    -1.    -1.    -1.    -1.    -1.    -1.    -1.    -1.
 -1.     0.323  1.     1.     1.     0.33  -1.    -1.    -1.    -1.    -1.
 -1.    -1.    -1.    -1.    -1.    -0.459  0.899  1.     1.     1.     0.442
 -1.    -1.    -1.    -1.    -1.    -1.    -1.    -1.    -1.    -0.378
  0.934  1.     1.     1.     0.994 -0.695 -1.    -1.    -1.    -1.    -1.
 -1.    -1.    -1.    -0.955  0.755  1.     1.     1.     1.     1.    -0.156
 -0.999 -1.    -1.    -1.    -1.    -1.    -1.    -1.    -0.057  1.     1.
  0.791  0.594  1.     1.     1.     0.335 -0.844 -1.    -1.    -1.    -1.
 -1.    -0.87   0.853  1.     1.     0.005 -0.976 -0.415  0.906  1.     1.
  0.673 -0.817 -1.    -1.    -1.    -1.    -0.591  1.     1.     0.378
 -0.976 -1.    -1.    -0.51   0.891  1.     1.     0.728 -0.789 -1.    -1.
 -1.     0.194  1.     1.    -0.434 -1.    -1.    -1.    -1.    -0.578
  0.851  1.     1.     0.38  -0.998

In [122]:
# Calc center of coords

for digit in (Digits):
    number = len([x for x in train_file if int(x[0]) == int(digit[0])])
    print (number)
    for i in range(256):
        digit[i+1] = sum([x[i+1] for x in train_file if x[0] == digit[0]])/number

224
121
101
79
86
55
90
64
92
88


In [123]:
print (Digits[8,:])

[  8.00000000e+00  -9.96847826e-01  -9.61369565e-01  -8.90250000e-01
  -7.60380435e-01  -5.31119565e-01  -1.95978261e-01   9.17065217e-02
   1.96565217e-01   1.32554348e-01  -6.98152174e-02  -3.51000000e-01
  -6.71358696e-01  -9.19336957e-01  -9.77445652e-01  -9.95597826e-01
  -1.00000000e+00  -9.62934783e-01  -8.71586957e-01  -7.30065217e-01
  -3.71130435e-01   8.31413043e-02   4.44673913e-01   5.20586957e-01
   4.71293478e-01   3.43434783e-01   2.87021739e-01   1.38119565e-01
  -7.08804348e-02  -6.06815217e-01  -8.91054348e-01  -9.58228261e-01
  -9.93782609e-01  -9.41967391e-01  -8.29239130e-01  -6.42195652e-01
  -1.86728261e-01   2.68945652e-01   3.50586957e-01   7.22717391e-02
  -1.41413043e-01  -2.88347826e-01  -8.77500000e-02   5.82065217e-02
   6.16847826e-02  -3.72826087e-01  -7.73608696e-01  -9.02804348e-01
  -9.78554348e-01  -9.59467391e-01  -7.93847826e-01  -5.67152174e-01
  -1.46315217e-01   3.50597826e-01   2.02923913e-01  -3.87673913e-01
  -6.34804348e-01  -5.75423913e-01

In [124]:
# Calc number of points belonging to a digit
for i in range (10):
    print (i,len([x for x in train_file if x[0] == i]))

0 224
1 121
2 101
3 79
4 86
5 55
6 90
7 64
8 92
9 88


In [125]:
# Calc radius of coords

for digit in (Digits):
    print (digit[0], max([np.sqrt(np.sum( (digit[1:]-x[1:])**2 )) for x in train_file if x[0] == digit[0]]))

0.0 16.2844421617
1.0 10.0472608381
2.0 13.7213961144
3.0 13.8712356145
4.0 14.0595973531
5.0 14.3394141919
6.0 13.7792663711
7.0 13.2656398998
8.0 14.7863503116
9.0 14.400636271


In [126]:
# Calc distances

for i in range(10):
    for j in range(10):
        dist[i,j] = np.sqrt(np.sum( (Digits[i,1:]-Digits[j,1:])**2 ))

In [143]:
# Show distances

# digit, smallest distance from its center to another
# all distances
for i in range(10):
    print (i, round(heapq.nsmallest(2, dist[i,:])[1]),1, round(max(dist[i,:]),1))
    print (dist[i,:])

0 7.0 1 14.1
[  0.          14.11632505   8.4676043    8.41851306  10.42238041
   7.1602956    8.07245358  11.2664355    9.35279618  11.16367646]
1 9.0 1 14.1
[ 14.11632505   0.           9.93729726  11.09284496  10.16774664
  10.70190271  10.3942642   10.18027021   9.8649275    9.35390531]
2 6.0 1 9.9
[ 8.4676043   9.93729726  0.          6.81556273  7.45825589  7.18893224
  7.26266021  7.78517696  6.27333601  8.1040591 ]
3 5.0 1 11.1
[  8.41851306  11.09284496   6.81556273   0.           8.90393736
   5.22130493   8.67915869   8.15744542   6.44789738   7.87878536]
4 6.0 1 10.4
[ 10.42238041  10.16774664   7.45825589   8.90393736   0.           8.0336445
   8.9923067    6.77080263   7.06294536   5.87749827]
5 5.0 1 10.7
[  7.1602956   10.70190271   7.18893224   5.22130493   8.0336445    0.
   6.67756065   8.58367972   6.86045913   8.15238098]
6 7.0 1 10.7
[  8.07245358  10.3942642    7.26266021   8.67915869   8.9923067
   6.67756065   0.          10.65956987   8.27917452  10.15101593]

## We see:
1. The digit $1$ has the significantly smallest radius with 10px.
That means that most "ones" are written more similar, i.e. rather covering the same pixels, as the other digits.
2. The other digits have radiuses between 13 ($7$) and 14.8 ($8$).
3. The digit $0$ has the biggest radius with 16.3. Thus, the "zeroes" are written rather differently.
4. Two pairs whose centers are closest together are the digits $3$ and $5$, as well as $9$ and $7$.
In both cases, the clouds have distances to each other of just 5px.
Thus, these digits are hardest to distinguish.
5. The two digits whose centers are farthest from each other and thus are easiest to distinguish are $0$ and $1$.
6. The maximal radius of any of these clouds is up to three times bigger than the distance between two clouds.
This renders this approach's reliability questionable; although a statistical analysis of the clouds and an inspection of their individual $\sigma$ is needed to support this assertion.